snap/integrity: new API

sespiros opened the pull request:
This is rebased on top of https://github.com/canonical/snapd/pull/14871.

Jira: https://warthogs.atlassian.net/browse/FR-9881

alfonsosanchezbeato commented on snap/integrity/integrity.go:
```diff
@@ -20,144 +20,90 @@
-// align aligns input `size` to closest `blockSize` value
-func align(size uint64) uint64 {
-	return (size + blockSize - 1) / blockSize * blockSize
+type IntegrityDataParams struct {
+	Type          string
+	Version       uint
+	HashAlg       string
+	DataBlocks    uint64
+	DataBlockSize uint64
+	HashBlockSize uint64
+	Digest        string
+	Salt          string
```
This type and the fields need a description

sespiros replied:
Added.

alfonsosanchezbeato commented on snap/integrity/integrity.go:
```diff
@@ -20,144 +20,90 @@
 
-	header := make([]byte, HeaderSize)
+	alg := strings.ReplaceAll(string(vsb.Algorithm[:]), "\x00", "")
+	encSalt := vsb.EncodedSalt()
```
I would put this nearer to usage

sespiros replied:
Thanks, I made the change.

alfonsosanchezbeato reviewed (commented):
Thanks, couple of minor comments

sespiros commented on snap/integrity/integrity.go:
```diff
@@ -20,144 +20,104 @@
+	// Check if the verity data that was found matches the passed parameters
+	alg := strings.ReplaceAll(string(vsb.Algorithm[:]), "\x00", "")
+	if alg != params.HashAlg {
+		return "", fmt.Errorf("snap integrity: dm-verity data %q were generated with an unasserted algorithm: %s != %s",
```
Yes that's a good point. What I had in mind is that this will be called after a step which retrieves the parameters from an assertion but as far as this API go, "unasserted" doesn't really make sense.

Looking at the PRs after this though, I think I would prefer if the error msg from snap-bootstrap was explicit that this was caused due to data that don't match the assertion/were unasserted. So there are 2 options to do this:
1) change unasserted to unexpected. Wrap these in a new error type. Filter the error type on the calling site to append/prepend context information.
2) split the verification in a separate function such as `VerifyData(integrityData asserts SnapIntegrityData) error`

Tthoughts?

pedronis commented:
> 1. change unasserted to unexpected. Wrap these in a new error type. Filter the error type on the calling site to append/prepend context information.
> 2. split the verification in a separate function such as VerifyData(integrityData asserts SnapIntegrityData) error

A 3rd approach is to pass the assertion to the function as it is but rename it as LookupVerifyDataAndCrossCheck ?

I'm fine kind of fine with any of approaches, it depends whether we have use cases where we do not want to cross-check, otherwise the last might be the simplest?

sespiros commented:
> > 1. change unasserted to unexpected. Wrap these in a new error type. Filter the error type on the calling site to append/prepend context information.
> > 2. split the verification in a separate function such as VerifyData(integrityData asserts SnapIntegrityData) error
> 
> A 3rd approach is to pass the assertion to the function as it is but rename it as LookupVerifyDataAndCrossCheck ?
> 
> I'm fine kind of fine with any of approaches, it depends whether we have use cases where we do not want to cross-check, otherwise the last might be the simplest?

@pedronis I pushed a commit:
1. stopped exporting `veritySuperBlock.Validate()` which will now be called as part of `ReadSuperBlock()`. Now all expected superblock validation errors will be thrown by `ReadSuperBlock()`.
1. renamed `LookupDmVerityData` to `LookupDmVerityDataAndCrossCheck`.

1. created 2 error types  `ErrDmVerityDataNotFound` and   `ErrUnexpectedDmVerityData`. 
    1. If `ErrDmVerityDataNotFound`, snap-bootstrap will use this to make a decision about whether to generate data (default behavior for install mode).
    1. If `ErrUnexpectedDmVerityData`, snap-bootstrap will wrap the error with extra context such as "dm-verity data found on disk don't match assertion".
    1. All other errors will be forwarded.

pedronis reviewed (commented):
thanks for the changes, some small comments

sespiros commented on snap/integrity/integrity.go:
```diff
@@ -20,144 +20,108 @@
 
-	copy(header, append(magic, jsonHeader...))
+var ErrDmVerityDataNotFound = errors.New("dm-verity data not found")
+var ErrUnexpectedDmVerityData = errors.New("unexpected dm-verity data")
```
Added.

sespiros commented on snap/integrity/integrity.go:
```diff
@@ -20,144 +20,108 @@
-	if firstNull == -1 {
-		return fmt.Errorf("invalid integrity data header: no null byte found at end of input")
+	if err != nil {
+		return "", err
```
I added an extra test case.

sespiros commented on snap/integrity/integrity.go:
```diff
@@ -20,144 +20,108 @@
-	Size     uint64 `json:"size,string"`
-	RootHash string `json:"dm-verity"`
-}
+func (params *IntegrityDataParams) check(vsb *dmverity.VeritySuperblock) error {
```
Made the change.

alfonsosanchezbeato reviewed (approved):
LGTM, thanks

pedronis commented on snap/integrity/integrity.go:
```diff
@@ -80,7 +80,11 @@ func (params *IntegrityDataParams) check(vsb *dmverity.VeritySuperblock) error {
+// ErrDmVerityDataNotFound is returned when dm-verity data for a snap are not found next to it.
 var ErrDmVerityDataNotFound = errors.New("dm-verity data not found")
+
+// ErrDmVerityDataNotFound is returned when dm-verity data for a snap are available but don't match
```
the name here seems a copy of the wrong one

sespiros replied:
Thanks for noticing

pedronis reviewed (approved):
thanks, small doc comment fix needed though

sespiros commented:
Hi @ernestl, it's weird I didn't hit this locally. I pushed a fix and will monitor the spread tests more closely.
